scrapy/pipelines/media.py

```diff
@@ -93,7 +94,7 @@ def _process_request(self, request, info, item):
         fp = self._fingerprinter.fingerprint(request)
         cb = request.callback or (lambda _: _)
         eb = request.errback
-        request.callback = None
```

Comment:
We need to update the `FilesPipeline` as well: https://github.com/scrapy/scrapy/blob/master/scrapy/pipelines/files.py#L495

Reply:
Since tests were actually passing for files, I had a deeper look at it, and I think tests pass because those `Request(u)` objects are later parsed with the very method here, so the callback gets set before the request object leaves the middleware. So I think no further changes may be necessary specific to files or images.

Reply:
Hm, it might still be cleaner to add the no callback marker to these requests, as they're not supposed to use "parse" callback.

Reply:
So we set callback to `NO_CALLBACK` twice, in `get_media_requests` and in `_process_request` (both called from `process_item`, one after the other)?

I am not against it, I just want to be certain that I made it clear enough that the reason the callback is not set here is because these request objects are processed further before they leave the pipeline, so with the current code there is no risk of anything outside the pipeline itself to receive a request with `callback=None`.

Reply:
> So we set callback to NO\_CALLBACK twice, in get\_media\_requests and in \_process\_request (both called from process\_item, one after the other)?

Yes. I think if the reader sees `FilesPipeline.get_media_requests()` with `Request(u, callback=NO_CALLBACK)`, it helps re-assure the idea that the `parse()` method isn't supposed to be involved here.

Although they could also further inspect `MediaPipeline._process_request()` and see that `NO_CALLBACK` is assigned, they won't have to if `FilesPipeline.get_media_requests()` already shows it.

Reply:
I’m making the change, then.

I wonder if we should go further, though, by changing `_process_request` to:
- Raise an exception if `callback` is anything other than `None` or `NO_CALLBACK`.
- Log a deprecation warning if `callback` is `None`.

Reply:
> Log a deprecation warning if `callback` is `None`.

+1 

> Raise an exception if `callback` is anything other than `None` or `NO_CALLBACK`. Or the same behavior as above, to avoid a backward-incompatible change. But I think it may be wise to actually break such code, to force users to not set a callback that is being reset in `_process_request`.

I'm not quite sure about this, since there might be some Scrapy project out there that does things differently with their MediaPipeline/FilePipeline. For example, they've overridden `_process_request` to not directly use the downloader.

Reply:
> But I think it may be wise to actually break such code, to force users to not set a callback that is being reset in _process_request.

The callback is actually not just reset, but stored and used. So maybe my point is void, we should continue to support callbacks on requests from `get_media_requests()` as usual. `_process_request` will make sure that the request leaves the pipeline with `callback=NO_CALLBACK`, but the original callback will be called nonetheless by the pipeline.